src/app/tests/TestReportScheduler.cpp

```diff
@@ -387,7 +457,351 @@ class TestReportScheduler
+        NL_TEST_ASSERT(aSuite, syncScheduler.mTestNextReportTimestamp == node3->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node1->GetSyncTimestamp() == node3->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node2->GetSyncTimestamp() == node3->GetMaxTimestamp());
+
+        sTestTimerSynchronizedDelegate.IncrementMockTimestamp(System::Clock::Milliseconds64(2000));
+
+        NL_TEST_ASSERT(aSuite, syncScheduler.IsReportableNow(readHandler1));
+        NL_TEST_ASSERT(aSuite, syncScheduler.IsReportableNow(readHandler2));
+        NL_TEST_ASSERT(aSuite, syncScheduler.IsReportableNow(readHandler3));
+        readHandler1->mObserver->OnBecameReportable(readHandler1);
+        readHandler2->mObserver->OnBecameReportable(readHandler2);
+        readHandler3->mObserver->OnBecameReportable(readHandler3);
+        // Engine run should happen here and send all reports
+        readHandler1->mObserver->OnSubscriptionAction(readHandler1);
+        readHandler2->mObserver->OnSubscriptionAction(readHandler2);
+        readHandler3->mObserver->OnSubscriptionAction(readHandler3);
+        NL_TEST_ASSERT(aSuite, !syncScheduler.IsReportableNow(readHandler1));
+        NL_TEST_ASSERT(aSuite, !syncScheduler.IsReportableNow(readHandler2));
+        NL_TEST_ASSERT(aSuite, !syncScheduler.IsReportableNow(readHandler3));
+        NL_TEST_ASSERT(aSuite, syncScheduler.mTestNextReportTimestamp == node1->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node1->GetSyncTimestamp() == node1->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node2->GetSyncTimestamp() == node1->GetMaxTimestamp());
+
+        // Now simulate a new readHandler being added with a max forcing a conflict
+        ReadHandler * readHandler4 =
+            readHandlerPool.CreateObject(nullCallback, exchangeCtx, ReadHandler::InteractionType::Subscribe);
+        NL_TEST_ASSERT(aSuite, CHIP_NO_ERROR == readHandler4->SetMaxReportingInterval(1));
+        NL_TEST_ASSERT(aSuite, CHIP_NO_ERROR == readHandler4->SetMinReportingIntervalForTests(0));
+        readHandler4->MoveToState(ReadHandler::HandlerState::GeneratingReports);
+        readHandler4->SetObserver(&syncScheduler);
+        readHandler4->mObserver->OnReadHandlerCreated(readHandler4);
+
+        // Confirm all handler are currently registered in the scheduler
+        NL_TEST_ASSERT(aSuite, syncScheduler.GetNumReadHandlers() == 4);
+        ReadHandlerNode * node4 = syncScheduler.FindReadHandlerNode(readHandler4);
+
+        // Confirm next report is scheduled on the max timestamp of readHandler4 and other handlers be synced
+        NL_TEST_ASSERT(aSuite, syncScheduler.mTestNextReportTimestamp == node4->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node1->GetSyncTimestamp() == node4->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node2->GetSyncTimestamp() == node4->GetMaxTimestamp());
+        NL_TEST_ASSERT(aSuite, node3->GetSyncTimestamp() == node1->GetMaxTimestamp());
```

Comment:
small nit. Even if node3 == node1 == node4 because of line 681 we should check node4 to be coherent.

Reply:
It has to be different since node1->GetMaxTimestamp != node4->GetMaxTimestamp. node3 can't sync on node4 since it is below its min, it will have a sync timestamp on a later point, which is in this case node1.